### PR TITLE
Superadmin Implementation

### PR DESCRIPTION
**Pull request by BrandonCookeDev**
This PR implements:
- Superadmin vs Region Admin privileges
- New UI on the Admin Functions that shows every user some information and allows them to change password

### CODE REVIEW

**Comment by BrandonCookeDev** [comment]
I still need to implement a unit test or two before we PR this

**Comment by BrandonCookeDev** [comment]
Tests written and this PR is ready to be reviewed

File: model.py [hunk 1]
```diff
@@ -348,7 +349,8 @@ class User(orm.Document):
               ('username', orm.StringField(required=True)),
               ('salt', orm.StringField(required=True)),
               ('hashed_password', orm.StringField(required=True)),
-              ('admin_regions', orm.ListField(orm.StringField()))]
+              ('admin_regions', orm.ListField(orm.StringField())),
+              ('admin_level', orm.StringField(required=True, default='REGION'))]
```
Comment (jschnei) [inline model.py]: since these guys should come from an Enum, probably it is a good idea to add a choices validator so the ORM will automatically check for this. (take a look at how tournament source types work on line 129).

File: model.py [hunk 2]
```diff
@@ -371,3 +373,8 @@ class Session(orm.Document):
     collection_name = 'sessions'
     fields = [('session_id', orm.StringField(required=True)),
               ('user_id', orm.StringField(required=True))]
+
+
+class AdminLevels(Enum):
```
Comment (jschnei) [inline model.py]: probably makes most sense to move this up to the top, next to the SOURCE_TYPE_CHOICES "enum".

For that matter, I'm still not sure if we gain anything by making this an explicit enum instead of just writing a tuple of choices (as in SOURCE_TYPE_CHOICES) and using the validator. It would make more sense I feel if we were mapping human-readable enum values to non-human readable int values (this is what a lot of enums are for), but since we are mapping them to string values, it feels a bit redundant (also, in server.py, it looks like we actually just use the explicit string constants).

Anyway tl;dr either way (enums/lists of strings) is okay, but if we decide to go the Enum route we should probably also convert SOURCE_TYPE_CHOICES to an Enum for consistency.

File: server.py [hunk 3]
```diff
@@ -137,6 +144,8 @@ def get_user_from_request(request, dao):
 
 
 def is_user_admin_for_region(user, region):
+    if user.admin_level == 'SUPER':
```
Comment (jschnei) [inline server.py]: if we want to stick with the Enum method, these 'SUPER's and everything should be M.AdminLevels.SUPERs.

File: server.py [hunk 4]
```diff
@@ -1176,8 +1209,8 @@ def put(self):
         user = get_user_from_request(request, dao)
         if not user:
             return 'Permission denied', 403
-        #if not is_user_admin_for_region(user, region='*'):
-        #    return 'Permission denied', 403
+        if not dao.get_is_superadmin(user.id):
```
Comment (jschnei) [inline server.py]: hmm if we have the user object, can't we just check if user.admin_level=='SUPER'? This should save a mongo lookup.

File: dao.py [hunk 5]
```diff
@@ -559,7 +559,7 @@ def insert_user(self, user):
 
     # throws invalidRegionsException, which is okay, as this is only used by a
```
Comment (jschnei) [inline dao.py]: let's remove this comment, it's no longer true with the introduction of the admin dashboard.

File: dao.py [hunk 6]
```diff
@@ -645,6 +665,17 @@ def check_creds_and_get_session_id_or_none(self, username, password):
         else:
             return None
 
+    def check_creds(self, username, password):
```
Comment (jschnei) [inline dao.py]: we should try not to repeat code (this is basically part of check_creds_and_get_session_id_or_none). maybe let's make `check_creds` return a user object, and refactor `check_creds_and_get_session_id_or_none` to call `check_creds` first?

(this could also be postponed to a later PR, but we should make a note).

File: dao.py [hunk 7]
```diff
@@ -619,6 +620,25 @@ def get_user_by_session_id_or_none(self, session_id):
...
 
+    def get_is_superadmin(self, user_id):
+        user = None
+        if self.users_col.find_one({'_id': user_id}):
+            user = M.User.load(self.users_col.find_one({'_id': user_id}))
+            return user.admin_level == 'SUPER'
+        else:
+            return False
+
+    def set_user_superadmin(self, user_id, admin_level):
```
Comment (jschnei) [inline dao.py]: this can be used for any admin_level, so let's call it something like `set_user_admin_level` or `update_user_admin_level`? (it might also make more sense to just make a general `update_user` function in the same vein as `update_player` and `update_tournament`).

File: server.py [hunk 8]
```diff
@@ -1196,9 +1229,14 @@ def put(self):
             uperm = args['new_user_permissions']
             uregions = args['new_user_regions']
 
+            perm = None
+            if uperm == 'Super Admin': perm = M.AdminLevels.SUPER
```
Comment (jschnei) [inline server.py]: for consistency, let's make the values of the permissions on the frontend match the string values of the enums on the backends (i.e. 'SUPER' and 'REGION')

File: webapp/app/tools/admin_functions/views/admin_functions.html [hunk 9]
```diff
@@ -1,65 +1,106 @@
...
+                      <form ng-submit="submitNewUser()" id="newUserForm">
+                        <h4>Add New Admin</h4>
+                        <p ng-model="userStatusMessage" id="userStatusMessage"></p>
+                        <label>Username:</label> <input type="text" id="new_user_name" class="form-control mobile" ng-model="postParams.new_user_name"> <br/>
+                        <label>Password:</label> <input type="password" id="new_user_pass" class="form-control" ng-model="postParams.new_user_pass"> <br/>
+                        <br/>
+
+                        <h4>Permission Level:</h4>
+                        <select ng-model="postParams.new_user_permissions" id="permissionLevelSelect">
+                            <option>Region Admin</option>
```
Comment (jschnei) [inline webapp/app/tools/admin_functions/views/admin_functions.html]: can we put value='REGION' and value='SUPER' here to pass 'REGION' and 'SUPER' to the backend? (not a big deal, but it's nice when values are consistent between frontend and backend)

File: server.py (hunk 3 above)
Comment (BrandonCookeDev) [inline server.py]: Going with the String method. so no need for this

File: model.py (hunk 2 above)
Comment (BrandonCookeDev) [inline model.py]: Going with the string method so no need for this

File: model.py (hunk 1 above)
Comment (BrandonCookeDev) [inline model.py]: So I've gone with the String Only method, but I'm not really sure how to validate on the ORM. can you point me in the right direction?

Comment (jschnei) [inline model.py]: so, take a look at line 129 to see how it's done for tournament source_types (challonge, tio, smashgg).

basically, you should do something like
```
              ('admin_level', orm.StringField(
                  required=True,
                  validators=[orm.validate_choices(ADMIN_LEVEL_CHOICES)]))
```

and define a list/tuple ADMIN_LEVEL_CHOICES somewhere.

File: dao.py (hunk 6 above)
Comment (BrandonCookeDev) [inline dao.py]: Oh looking a little closer it might be because it finds no user in the db and then returns None...

File: dao.py [hunk 10]
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
...
+    def get_is_superadmin(self, user_id):
+        user = None
+        if self.users_col.find_one({'_id': user_id}):
+            user = M.User.load(self.users_col.find_one({'_id': user_id}))
+            return user.admin_level == 'SUPER'
+        else:
+            return False
+
+    def set_user_admin_level(self, user_id, admin_level):
+        if type(admin_level) is not M.AdminLevels:
```
Comment (jschnei) [inline dao.py]: I think M.AdminLevels does not exist anymore, so this should error (is it being tested)?
In any case, the ORM should take care of this check.

File: dao.py [hunk 11]
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
     def get_user_by_region(self, regions):
         pass
 
+    def get_is_superadmin(self, user_id):
+        user = None
```
Comment (jschnei) [inline dao.py]: I think it is probably better to use `get_user_by_id_or_none` here instead of rewriting the Mongo queries (so something like `user = get_user_by_id_or_none(user_id)`.

Actually, do we use this function (get_is_super_admin) anywhere else in the code? If not, I think it's fine to remove this function (it is easy enough to just check whether user.admin_level=='SUPER' whenever we have a user object).

File: dao.py [hunk 12]
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
...
+        if self.users_col.find_one({'_id': user_id}):
+            user = M.User.load(self.users_col.find_one({'_id': user_id}))
+            return user.admin_level == 'SUPER'
+        else:
+            return False
+
+    def set_user_admin_level(self, user_id, admin_level):
```
Comment (jschnei) [inline dao.py]: do we actually use this function anywhere else in the code yet? if not, I think it would be fine to just replace it with a simple `update_user` function, in the same style as `update_player` and `update_region`.

File: dao.py [hunk 13]
```diff
@@ -619,6 +618,25 @@ def get_user_by_session_id_or_none(self, session_id):
...
+        if type(admin_level) is not M.AdminLevels:
+            raise Exception('Submitted admin level is not of correct type')
+        if self.users_col.find_one({'_id': user_id}):
```
Comment (jschnei) [inline dao.py]: right now this way of updating bypasses the ORM completely and won't trigger any of the ORMs validators. the correct way to do this sort of thing is 1. read the user data into a User object (as is done in get_user_by_id_or_none), 2. update whatever properties of the User object, 3. then write the User object back into the database using `self.users_col.update({'_id': user.id}, user.dump(context='db'))`.

File: dao.py (hunk 6 above)
Comment (jschnei) [inline dao.py]: i'm confused, it looks there are now 3 functions which repeat a lot of code?:

1. `check_creds(username, password)` which returns whether (username, password) are valid credentials.
2. `check_creds_and_get_session_id_or_none(username, password)`, which checks if (username, password) are valid credentials, and tries to return the current session
3. `get_and_verify_user_by_username`, which seems to be a duplicate of `get_user_by_username`.

Right now 1 and 2 duplicate a lot of the same code. I recommend removing `get_and_verify_user_by_username` and refactoring 1 and 2 so that check_creds returns a user object (not just a bool), and so that get_session_id calls check_creds to get the user object.

**Comment by BrandonCookeDev** [comment]
Jon, I made the changes. Notably I deleted the third check_creds thing I had written. I left it in to try to fix whatever bug was in it, but now I have no time.
Also I rewrote the set_admin_level function, and commented it out. Since it isn't implemented yet, and we may want it in the future, I felt this was the best course of action

**Comment by jschnei** [comment]
I fixed a bug with creating new users; I think it works now.

Also, while fixing this I noticed some of the frontend code is probably less angular-y than it should be; there's a lot of places where you're manipulating the DOM directly (I think the angular philosophy is that this should basically always be done through angular directives). Fixing this can probably wait for a later PR though; there is a bunch of frontend refactoring that still needs to be done.
